### PR TITLE
Optimizations

### PR DESCRIPTION
**Pull request by babenek**
## Description

Please include a summary of the change and which is fixed.

- Add cached property for Target and optimized using the properties
- Use set() for rules. required_substrings to avoid extra transformations in future

## How has this been tested?

Please describe the tests that you ran to verify your changes.

- [x] UnitTest
- [x] Benchmark

### CODE REVIEW

File: credsweeper/scanner/scanner.py [hunk 1]
```diff
@@ -43,6 +43,21 @@ def __init__(self, config: Config, rule_path: Union[None, str, Path]) -> None:
         self.min_len = min(self.min_pattern_len, self.min_keyword_len, self.min_pem_key_len, self.min_multi_len,
                            MIN_VARIABLE_LENGTH + MIN_SEPARATOR_LENGTH + MIN_VALUE_LENGTH)
 
+    def _get_required_substrings(self, rule_type: RuleType) -> Set[str]:
+        """init set of required substrings for custom rule type"""
+        required_substrings: Set[str] = set()
+        for rule in (x[0] for x in self.rules_scanners if rule_type == x[0].rule_type):
+            required_substrings.update(set(rule.required_substrings))
+        return required_substrings
+
```
Comment (babenek) [inline credsweeper/scanner/scanner.py:52]:
```suggestion
```

File: credsweeper/filters/value_ip_check.py [hunk 2]
```diff
@@ -32,9 +32,10 @@ def run(self, line_data: LineData, target: AnalysisTarget) -> bool:
         with contextlib.suppress(Exception):
             ip = ipaddress.ip_address(line_data.value)
             if 4 == ip.version:
-                line_lower = target.line.lower()
+                # use line_strip_lower due the property should be cached already
+                line_strip_lower = target.line_strip_lower
                 for i in ValueIPCheck.FALSE_POSITIVE_MARKERS:
-                    if i in line_lower:
+                    if i in line_strip_lower:
```
Comment (csh519) [inline credsweeper/filters/value_ip_check.py:38]: Is it okay to use `target.line_strip` instead of `target.line`?

Comment (babenek) [reply credsweeper/filters/value_ip_check.py:38]: Yes, it is OK. 
Shorter string - quickly process. Whitespaces are not necessary to research

Comment (csh519) [reply credsweeper/filters/value_ip_check.py:38]: I see. :)

**Review by csh519 — APPROVED** [review]
Thank you for the optimizations.

LGTM :+1:

**Review by xDizzix — APPROVED** [review]
+

**Pull request merged**